### PR TITLE
Update Doc-String of MapDriver.get_iter

### LINKED ISSUE #60: [FEATURE] Make `get_iter` method documentation about `max_workers` more explicit
Hey, I've stumbled across a potentially easy-to-misunderstand part of the `MapDriver.get_iter` documentation:

https://github.com/merantix-momentum/squirrel-core/blob/8e2942313c7d7dd974b1ca2f2308895f660d3d26/squirrel/driver/driver.py#L68-L155

The documentation of `max_workers` states that by default `None` will be used and also _mentions_ that this will cause `async_map` to be called but I missed these parts of the documentation and was surprised to see that so many threads were allocated.

I am/was not too familiar with the `ThreadPoolExecutor` interface and find it somewhat surprising that `None` equals `numer_of_processors x 5` according to the [ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor) definition.
Maybe it would be helpful to explicitly state that by default ThreadPoolExecutor will be used with so many threads?
The documentation string reads a bit unintuitive as the starts out that max_worker defines how many items are fetched simultaneously with `max_worker` and then continues to state that _otherwise_ `map` is used.
From that perspective, `max_workers=None` doesn't sound like it should be using any threads at all.
Without knowing the default values of `ThreadPoolExecutor` I would make it more explicit that to _disable_ threading one has to set max_workers=0/1 and that by default _many_ threads are used.

I am happy to add a PR with my suggested doc-string update if you agree! :)

### PR DESCRIPTION
**Pull request by kai-tub**
- Better document the behavior of `max_workers` and link to official `ThreadPoolExecutor` documentation.
- Update `*_map` doc-strings that use `ThreadPoolExecutor` and link to official `ThreadPoolExecutor` documentation.

Fixes #60  issue

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation update
- [ ] Refactoring including code style reformatting 
- [ ] Other (please describe):

# Checklist:

- [X] I have read the [contributing guideline doc](https://squirrel-core.readthedocs.io/en/latest/usage/code_of_conduct.html) (external contributors only)
- [ ] Lint and unit tests pass locally with my changes
- [X] I have kept the PR small so that it can be easily reviewed  
- [X] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] All dependency changes have been reflected in the pip requirement files.

### CODE REVIEW

**Comment by winfried-loetzsch** [comment]
Hi @kai-tub thanks for your contribution! This looks good already

File: squirrel/driver/driver.py [hunk 1]
```diff
@@ -98,11 +98,17 @@ def get_iter(
...
+                is called to fetch the items iteratively.
+                If `max_workers` is bigger than 1 or equal to None,
+                :py:meth:`~squirrel.iterstream.Composable.async_map` is called to fetch multiple items simultaneously.
+                In this case, the `max_workers` argument refers to the maximum number of
+                workers in the :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>`
+                Pool of :py:meth:`~squirrel.iterstream.Composable.async_map`.
+                `None` has a special meaning in this context and uses an internal heuristicZ for the number of workers.
+                The exact number of workers with `max_workers=None` depends on the specific Python version.
+                See :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>` for details.
+                Defaults to None.rea
```
Comment (winfried-loetzsch) [inline squirrel/driver/driver.py:111]:
```suggestion
                Defaults to None.
```

File: squirrel/driver/driver.py [hunk 2]
```diff
@@ -98,11 +98,17 @@ def get_iter(
...
-                Otherwise, :py:meth:`~squirrel.iterstream.Composable.map` is called and `max_workers` is not used.
-                Defaults to None.
+            max_workers (int, optional): If `max_workers`  is equal to 0 or 1, :py:meth:`~squirrel.iterstream.Composable.map`
+                is called to fetch the items iteratively.
+                If `max_workers` is bigger than 1 or equal to None,
+                :py:meth:`~squirrel.iterstream.Composable.async_map` is called to fetch multiple items simultaneously.
+                In this case, the `max_workers` argument refers to the maximum number of
+                workers in the :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>`
+                Pool of :py:meth:`~squirrel.iterstream.Composable.async_map`.
+                `None` has a special meaning in this context and uses an internal heuristicZ for the number of workers.
```
Comment (winfried-loetzsch) [inline squirrel/driver/driver.py:108]:
```suggestion
                `None` has a special meaning in this context and uses an internal heuristic for the number of workers.
```

**Review by winfried-loetzsch — CHANGES_REQUESTED** [review]
Just a few typos, otherwise looks good!

File: docs/usage/Iterstream.rst [hunk 3]
```diff
@@ -86,14 +87,23 @@ Part of the fast speed from iterstream thanks to :py:meth:`squirrel.iterstream.b
         time.sleep(1)
         return item
 
-    it = IterableSource([1, 2, 3]).async_map(io_bound, executor=tpool).async_map(io_bound)
+    it = IterableSource([1, 2, 3]).async_map(io_bound, executor=tpool, max_workers=4).async_map(io_bound)
```
Comment (AlpAribal) [inline docs/usage/Iterstream.rst:90]: I suggest keeping `max_workers` here, removing `tpool` and explaining how to specify an `executor` below the code block (which you have already done)
```suggestion
    it = IterableSource([1, 2, 3]).async_map(io_bound, max_workers=4).async_map(io_bound)
```

File: squirrel/driver/driver.py (hunk 1 above)
Comment (AlirezaSohofi) [inline squirrel/driver/driver.py:111]:
```suggestion
                Defaults to None.
```

File: squirrel/driver/driver.py (hunk 2 above)
Comment (AlirezaSohofi) [inline squirrel/driver/driver.py:108]:
```suggestion
                `None` has a special meaning in this context and uses an internal heuristic of :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>` for the number of workers.
```

File: docs/usage/Iterstream.rst [hunk 4]
```diff
@@ -86,14 +87,23 @@ Part of the fast speed from iterstream thanks to :py:meth:`squirrel.iterstream.b
...
-`async_map` instantiates a :code:`concurrent.futures.ThreadPoolExecutor` if the argument `executor` is `None` (default). It also accepts :code:`concurrent.futures.ProcessPoolExecutor`, which is a good choice when performing cpu-bound operations on a single machine.
+:py:meth:`async_map <squirrel.iterstream.base.Composable.async_map>`
+instantiates a :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>`
+if the argument `executor` is `None` (default).
+It also accepts :py:class:`ProcessPoolExecutor <concurrent.futures.ProcessPoolExecutor>`,
+which is a good choice when performing cpu-bound operations on a single machine.
 
+The argument `max_workers` defines the maximum number of workers of the selected
+`executor`. By default, `max_workers=None` relies on an internal heuristic of
+the `executor` to select a reasonable upper bound.
```
Comment (AlpAribal) [inline docs/usage/Iterstream.rst:104]: We should make explicit when exactly `max_workers` is being used.
```suggestion
If an executor is not provided, the argument `max_workers` can be used to control the level of parallelism.
This argument defines the maximum number of workers used by the ThreadPoolExecutor
that will be instantiated by squirrel.
By default, `max_workers=None` relies on an internal heuristic of
ThreadPoolExecutor to select a reasonable upper bound.
```

**Review by AlirezaSohofi — CHANGES_REQUESTED** [review]
Thanks looks very good, left some minor comments :+1:

File: docs/usage/Iterstream.rst [hunk 5]
```diff
@@ -86,14 +87,23 @@ Part of the fast speed from iterstream thanks to :py:meth:`squirrel.iterstream.b
...
+This may differ between Python versions.
+See the documentation of the `executor` for more details.
```
Comment (AlpAribal) [inline docs/usage/Iterstream.rst:106]:
```suggestion
See the documentation of the `executor` for more details.
If an `executor` is specified, then `max_workers` is ignored.
```

File: squirrel/iterstream/base.py [hunk 6]
```diff
@@ -158,10 +158,12 @@ def async_map(
         Args:
             callback (Callable): a callable to be applied to items in the stream
             buffer (int): the size of the buffer
-            max_workers (int): number of workers in the ThreadPoolExecutor
+            max_workers (int): number of workers in the :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>`
```
Comment (AlpAribal) [inline squirrel/iterstream/base.py:161]:
```suggestion
            max_workers (int): number of workers in the :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>`
                `max_workers` is only used when `executor` is not provided.
```

**Review by AlpAribal — CHANGES_REQUESTED** [review]
Thank you for the PR and also for the smaller improvements here and there!

Actually, if an executor is provided, `max_workers` is not used (since the number of workers was already set when instantiating the executor). This is currently not mentioned in the docs and it would be nice to add this information as well.

I have made a few suggestions on how/where to include this info. These are only suggestions though, feel free to use your own words/structure.

**Comment by kai-tub** [comment]
Ah, I completely missed that max_workers is only applied if `executor` is not changed.
I will need some time to update the documentation in various places then.
Once I am done, I will remove `[WIP]` from the title again.

File: docs/usage/Iterstream.rst (hunk 3 above)
Comment (kai-tub) [reply docs/usage/Iterstream.rst:90]: :+1: I will add this to the documentation

**Comment by AlpAribal** [comment]
> Wouldn't it be "more obvious" to call `max_workers` `max_threads` instead?

In hindsight, yes, it sounds more intuitive to me. On the other hand, `ThreadPoolExecutor` also calls its "n_threads" argument `max_workers`, so maybe using the same convention is better. I do not have a strong feeling regarding this. I would wait a while before changing the name, and try to validate if the naming does create more and more confusion among users.

**Comment by kai-tub** [comment]
A third option would be to add a simple check if `max_workers is not None and executor is not None` and raise a UserWarning that the `max_workers` argument will be ignored.

**Comment by kai-tub** [comment]
I have read the CLA Document and I hereby sign the CLA

File: squirrel/driver/driver.py [hunk 7]
```diff
@@ -98,10 +98,16 @@ def get_iter(
                     depending on whether `func` is a subclass of :py:meth:`~squirrel.iterstream.Composable` or a
                     `Callable`, one of the above cases will happen, with the only difference that arguments are passed
                     too. This is useful for passing arguments.
-            max_workers (int, Optional): If larger than 1 or None, :py:meth:`~squirrel.iterstream.Composable.async_map`
-                is called to fetch multiple items simultaneously and `max_workers` refers to the maximum number of
-                workers in the ThreadPoolExecutor used by `async_map`.
-                Otherwise, :py:meth:`~squirrel.iterstream.Composable.map` is called and `max_workers` is not used.
+            max_workers (int, optional): If `max_workers`  is equal to 0 or 1, :py:meth:`~squirrel.iterstream.Composable.map`
```
Comment (AlpAribal) [inline squirrel/driver/driver.py:101]:
```suggestion
            max_workers (int, optional): If `max_workers` is equal to 0 or 1, :py:meth:`~squirrel.iterstream.Composable.map`
```

File: squirrel/iterstream/base.py [hunk 8]
```diff
@@ -158,14 +158,21 @@ def async_map(
...
```
Comment (AlpAribal) [inline squirrel/iterstream/base.py:161]:
```suggestion
            max_workers (int): number of workers in the :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>`.
```

File: docs/usage/Iterstream.rst [hunk 9]
```diff
@@ -73,26 +73,76 @@ Note that you can pass the probabilities of sampling from each iterator. When an
...
+:py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>`
+uses when `exectuor=None`.
+By default, `max_workers=None` relies on an internal heuristic of
+the :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>`
+to select a reasonable upper bound.
+This may differ between Python versions.
+See the documentation of
+:py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>` for details.
+
+In the above example, two :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>`'s
```
Comment (AlpAribal) [inline docs/usage/Iterstream.rst:108]:
```suggestion
In the above example, two :py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>`\s
```

**Review by AlpAribal — CHANGES_REQUESTED** [review]
Very nice examples and explanations, thank you! I only left a few minor comments.

**Comment by kai-tub** [comment]
Sorry for those typos...
Do you have any recommendations on how one could catch them automatically?
Can you somehow export the document as a plain text file and run a grammar-checker on it?
I am using VScode if you happen to know a good extension.
Thanks :)

**Comment by AlpAribal** [comment]
> Sorry for those typos... Do you have any recommendations on how one could catch them automatically? Can you somehow export the document as a plain text file and run a grammar-checker on it? I am using VScode if you happen to know a good extension. Thanks :)

I am using Code Spell Checker (extension id: streetsidesoftware.code-spell-checker). Just personal preference though, there are probably a bunch of other options as well.

**Comment by kai-tub** [comment]
Ah yes.
Merge conflicts... The arch-nemesis of junior developers...

**Comment by kai-tub** [comment]
I think I might have overwritten the previous changes in iterstream.
I will have a look at it again once I have some left-over time.
The changed name made it kinda hard to follow what changed between commits.
Please do not merge without reviewing it.

File: docs/usage/iterstream.rst [hunk 10]
```diff
@@ -2,15 +2,7 @@ IterStream
 ==========
 
 Squirrel provides an API for chaining iterables.
-The functionality is provided through the :py:class:`Composable` class, which acts as a base class for most classes in IterStream.
-
-Stream Processing Methods
```
Comment (pzdkn) [inline docs/usage/iterstream.rst:7]: @kai-tub , I think this shouldn't be overwritten

File: docs/usage/iterstream.rst [hunk 11]
```diff
@@ -64,8 +56,6 @@ Items in `IterableSource` can be composed by providing a Composable in the `comp
         print(item)
 
 
-To see how you can chain custom Composables with `compose()`, see the advanced section for :ref:`IterStream <advanced/iterstream:IterStream>`.
```
Comment (pzdkn) [inline docs/usage/iterstream.rst:67]: This should also remain

File: docs/usage/iterstream.rst [hunk 12]
```diff
@@ -83,25 +73,238 @@ Note that you can pass the probabilities of sampling from each iterator. When an
...
+In the above example, a
+:py:class:`ThreadPoolExecutor <concurrent.futures.ThreadPoolExecutor>` is created with
+a maximum of 4 workers.
+This pool of workers is shared among both
+:py:meth:`async_map <squirrel.iterstream.base.Composable.async_map>` calls.
+After exhausting the iterator, the `tpool` is shutdown.
+
+
+
+Cluster-mode
```
Comment (pzdkn) [inline docs/usage/iterstream.rst:148]: This has been moved to the integration section

File: docs/usage/iterstream.rst [hunk 13]
```diff
@@ -83,25 +73,238 @@ Note that you can pass the probabilities of sampling from each iterator. When an
...
+
+    it = IterableSource([1, 2, 3]).dask_map(io_bound).dask_map(lambda item: item + 1).materialize_dask()
+    t1 = time.time()
+    for item in it:
+        print(item)
+    print(time.time() - t1)
+
+Note that after calling :code:`dask_map` for the first time, you can chain more :code:`dask_map`\'s, which are then operating on the :code:`dask.delayed.Delayed` objects, so that the data and the operations live on the dask cluster until :code:`materialize_dask` is called.
+
+Just-in-time compilation with numba
```
Comment (pzdkn) [inline docs/usage/iterstream.rst:175]: This has been also moved

File: docs/usage/iterstream.rst [hunk 14]
```diff
@@ -83,25 +73,238 @@ Note that you can pass the probabilities of sampling from each iterator. When an
...
+.. code-block:: python
+
+    from squirrel.driver import MessagepackDriver
+    url = ""
+    it = MessagepackDriver(url).get_iter(key_hooks=[SplitByWorker]).async_map(times_two).batched(batch_size).compose(TorchIterable)
+    dl = DataLoader(it, num_workers=num_workers)
+
+In this example, :code:`key_hooks=[SplitByWorker]` ensures that keys are split between workers before fetching the data and we achieve two level of parallelism; multi-processing provided by :code:`torch.utils.data.DataLoader`, and multi-threading inside each process for efficiently fetching samples by :code:`get_iter`.
+
+Performance Monitoring
```
Comment (pzdkn) [inline docs/usage/iterstream.rst:252]: Also moved

**Review by pzdkn — CHANGES_REQUESTED** [review]
Thanks @kai-tub for your contribution :) It seems you some merge-conflict, I left some comment on what was overwritten and what was unintentionally moved.

File: docs/usage/iterstream.rst (hunk 13 above)
Comment (kai-tub) [reply docs/usage/iterstream.rst:175]: I cannot find the section where this has been moved.
It seem like the jit numba discussion has been removed altogether?

**Comment by kai-tub** [comment]
I hope that the PR is now finally ready to be merged. :)
I would like to thank everyone who was involved in the process for the patience to resolve the merge conflict and for fixing quite a few typos in my commits.
Please feel free to squash the commits and re-write the commit message however you like.
I feel like the number of commits from this PR would pollute the commit history of other files without having any clear benefit.

Thank you all again for your help!

**Review by pzdkn — APPROVED** [review]
LGTM
